Hide imported names in stubs unless 'as id' is used (PEP 484)

ilevkivskyi opened the pull request:
Fixes #2927 

PEP 484 specifies that only names imported as ``from mod import name as name`` should be re-exported. However, mypy didn't follow this rule so that this code passed without errors but obviously fails at runtime:
```python
from collections import TypeVar
from weakref import Generic
from getopt import List

T = TypeVar('T')

class C(Generic[T], List[int]):
    ...
```
This PR makes all these errors. CI will fail since there are several dozens of errors in typeshed, I am going to make a PR to typeshed soon.

gvanrossum commented:
I'll test this against Dropbox codebases.

ilevkivskyi commented:
It looks like some of the typeshed failures are due to a bug in my PR, will investigate now.

gvanrossum commented:
Yeah, I get this for example:
```
mypy -2 -c 'import xml.sax'
typeshed/stdlib/2and3/xml/sax/__init__.pyi:27: error: Name 'xml.sax.xmlreader.XMLReader' is not defined
typeshed/stdlib/2and3/xml/sax/__init__.pyi:29: error: Name 'xml.sax.handler.ContentHandler' is not defined
typeshed/stdlib/2and3/xml/sax/__init__.pyi:30: error: Name 'xml.sax.handler.ErrorHandler' is not defined
typeshed/stdlib/2and3/xml/sax/__init__.pyi:32: error: Name 'xml.sax.handler.ContentHandler' is not defined
typeshed/stdlib/2and3/xml/sax/__init__.pyi:33: error: Name 'xml.sax.handler.ErrorHandler' is not defined
typeshed/stdlib/2and3/xml/sax/__init__.pyi:35: error: Name 'xml.sax.xmlreader.XMLReader' is not defined
```

ilevkivskyi commented:
After I fixed the bug in this PR, it looks like there are only _two_ errors in typeshed:
```
error: Module 'json' has no attribute 'JSONDecodeError'
error: Module 'asyncio.futures' has no attribute 'Awaitable'
```
Both seem to be legitimate errors.

ilevkivskyi commented:
> only _two_ errors in typeshed

Actually two in stdlib plus a dozen or so in third party stubs.

gvanrossum commented:
So do we need to fix all the typeshed errors first? Also I can't re-test this with the Dropbox codebases until you've fixed the merge conflict.

ilevkivskyi commented:
@gvanrossum I fixed the merge conflict and made a PR to typeshed https://github.com/python/typeshed/pull/1484

gvanrossum commented:
**Please hold off on this until after the release of 0.521 (or until Jukka vets it).**

ilevkivskyi commented:
> Please hold off on this until after the release of 0.521 (or until Jukka vets it).

NP, I don't think this is urgent. But it would be great to test this with internal codebases (maybe there are more errors in stubs).

ilevkivskyi commented:
OK, I have fixed the logic as discussed in https://github.com/python/typeshed/issues/1484, this should be now ready for review and testing with internal codebases.

ilevkivskyi commented:
I added one more test and simplified the PR, still I cannot do this with only one flag ``module_public`` since it is used for other things like ``__all__``, ``import *`` and ``_private_names``. It is possible to have a single flag with three values ``HIDDEN, PRIVATE, PUBLIC = 0, 1, 2`` instead of two ``bool`` flags ``module_public`` and ``module_hidden`` but this will induce some code churn.

Also typeshed is synced, so that this ready to be reviewed and merged.

ethanhs commented on mypy/semanal.py:
```diff
@@ -3508,7 +3510,9 @@ def lookup_qualified(self, name: str, ctx: Context,
-            return n
+                    if n and n.module_hidden:
+                        self.name_not_defined(name, ctx)
+            return n if n and not n.module_hidden else None
```
Minor style nit. Take it if you want. I prefer to not have compound if statements in returns. It makes things clearer if you break it up, but up to you.

ilevkivskyi commented:
@gvanrossum Do you think Ethan's review is enough, or we need another one? (This is quite a minor change, so that I would merge this soon to avoid more merge conflicts.)

gvanrossum commented:
I'm taking a look now.
